File: leetcode/120-triangle.cpp
class Solution
{
 public:
  int minimumTotal(vector<vector<int> >& triangle)
  {
    int n = triangle.size();
    if (!n) return 0;
    const int imax = numeric_limits<int>::max();

    vector<int> sum(n, imax);
    sum[0] = 0;
    for (int i = 0; i < n; ++i) {
      int pre = imax;
      for (int j = 0; j <= i; ++j) {
        int tmp = sum[j];
        sum[j] = min(pre, sum[j]) + triangle[i][j];
        pre = tmp;
      }
    }
    return *min_element(sum.begin(), sum.end());
  }
};

File: leetcode/120-triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        n = len(triangle)
        if 0 == n: return 0

        imax = sys.maxsize
        sums = [0] + [imax] * (n-1)
        for i in range(n):
            p = imax
            for j in range(i+1):
                t = sums[j]
                sums[j] = min(p, sums[j]) + triangle[i][j]
                p = t
        return min(sums)
